File: packages/sdk/src/verification-compliance.ts
/** Typed SDK for Domain 73 — Verification, Compliance, and Identity Review Dashboard. */

export type VcRole       = 'viewer' | 'vc_analyst' | 'vc_lead' | 'vc_admin';
export type CaseStatus   = 'pending' | 'reviewing' | 'holding' | 'approved' | 'rejected' | 'escalated' | 'expired' | 'archived';
export type CaseQueue    = 'triage' | 'review' | 'escalation' | 'closed';
export type Program      = 'kyc' | 'kyb' | 'aml' | 'sanctions' | 'address' | 'tax' | 'accreditation' | 'right_to_work' | 'professional_licence';
export type SubjectKind  = 'user' | 'professional' | 'enterprise' | 'agency';
export type DocKind      = 'passport' | 'national_id' | 'driving_licence' | 'utility_bill' | 'bank_statement' | 'selfie' | 'company_reg' | 'tax_id' | 'licence' | 'other';
export type CheckType    = 'document' | 'facial_similarity' | 'watchlist' | 'pep' | 'sanctions' | 'address' | 'company_reg' | 'aml' | 'tax' | 'adverse_media';
export type CheckResult  = 'pending' | 'clear' | 'consider' | 'rejected' | 'error';
export type Decision     = 'approve' | 'reject' | 'request_more_info' | 'step_up' | 'hold' | 'escalate' | 'dismiss' | 'expire' | 'renew';
export type Severity     = 'low' | 'normal' | 'high' | 'critical';
export type Band         = 'normal' | 'elevated' | 'high' | 'critical';

export interface VcFlag { code: string; severity: Severity; source: string }

export interface VcCase {
  id: string; reference: string;
  subject_id: string; subject_kind: SubjectKind;
  program: Program; jurisdiction: string;
  risk_score: number; risk_band: Band;
  status: CaseStatus; queue: CaseQueue;
  assigned_to?: string | null;
  sla_due_at?: string | null; expires_at?: string | null;
  reasons: string[]; flags: VcFlag[];
  meta: Record<string, unknown>;
  created_at: string; updated_at: string;
}
export interface VcDocument {
  id: string; case_id: string; kind: DocKind;
  filename: string; storage_url: string;
  mime_type?: string | null; bytes?: number | null;
  hash_sha256?: string | null; ocr_text?: string | null;
  ocr_fields: Record<string, unknown>;
  liveness_score?: number | null; match_score?: number | null;
  status: 'pending' | 'accepted' | 'rejected' | 'expired';
  uploaded_at: string;
  reviewed_by?: string | null; reviewed_at?: string | null;
}
export interface VcCheck {
  id: string; case_id: string; provider: string;
  check_type: CheckType; result: CheckResult; score?: number | null;
  payload: Record<string, unknown>; external_id?: string | null;
  created_at: string; completed_at?: string | null;
}
export interface VcDecision {
  id: string; case_id: string; actor_id: string;
  decision: Decision; rationale: string;
  duration_days?: number | null;
  appealable: 'yes' | 'no'; meta: Record<string, unknown>;
  created_at: string;
}
export interface VcEvent {
  id: string; case_id?: string | null; subject_id?: string | null;
  actor_id?: string | null; action: string;
  from_state?: string | null; to_state?: string | null;
  diff: Record<string, unknown>; ip?: string | null; created_at: string;
}
export interface VcWatchlistEntry {
  id: string; subject_id: string; subject_kind: SubjectKind;
  reason: string; severity: Severity;
  added_by?: string | null; expires_at?: string | null;
  meta: Record<string, unknown>; created_at: string;
}
export interface VcKpis {
  casesByStatus:  Partial<Record<CaseStatus, number>>;
  casesByQueue:   Partial<Record<CaseQueue, number>>;
  casesByBand:    Partial<Record<Band, number>>;
  casesByProgram: Partial<Record<Program, number>>;
  slaBreached:    number;
  expiringSoon:   number;
  watchlist:      number;
}
export interface VcInsight  { id: string; severity: 'success' | 'info' | 'warn' | 'critical'; title: string }
export interface VcDeskRisk { score: number; band: Band; model: string; factors: Record<string, number> }
export interface VcOverview {
  kpis: VcKpis;
  queues: { triage: VcCase[]; review: VcCase[]; escalation: VcCase[] };
  watchlist: VcWatchlistEntry[];
  insights: VcInsight[]; deskRisk: VcDeskRisk; computedAt: string;
}
export interface VcCaseDetail {
  case: VcCase; documents: VcDocument[]; checks: VcCheck[];
  decisions: VcDecision[]; events: VcEvent[];
}
export interface VcList<T> { items: T[]; total: number; meta: { source: string; role: VcRole; page: number; pageSize: number } }

File: apps/ml-python/app/verification_compliance.py
"""Domain 73 — Verification & Compliance scoring (deterministic, explainable)."""
from __future__ import annotations
from typing import Any, Dict, List

from fastapi import APIRouter
from pydantic import BaseModel, Field

router = APIRouter(prefix="/verification-compliance", tags=["verification-compliance"])

HIGH_RISK_JUR = {"IR", "KP", "SY", "CU", "RU", "BY", "MM", "VE"}


class CaseIn(BaseModel):
    subjectId: str | None = None
    subjectKind: str = "user"
    program: str = "kyc"
    jurisdiction: str = "GB"
    reasons: List[str] = Field(default_factory=list)
    meta: Dict[str, Any] = Field(default_factory=dict)


class CheckIn(BaseModel):
    caseId: str | None = None
    provider: str = "manual"
    checkType: str = "document"


class DeskRiskIn(BaseModel):
    kpis: Dict[str, Any] = Field(default_factory=dict)


class InsightsIn(BaseModel):
    signals: Dict[str, Any] = Field(default_factory=dict)


def _band(score: int) -> str:
    if score >= 80: return "critical"
    if score >= 60: return "high"
    if score >= 35: return "elevated"
    return "normal"


@router.post("/score-case")
def score_case(body: CaseIn) -> Dict[str, Any]:
    flags: List[Dict[str, Any]] = []
    reasons: List[str] = []
    score = 10
    if body.program in {"aml", "sanctions", "pep"}:
        score += 40
        flags.append({"code": body.program, "severity": "critical", "source": "program"})
        reasons.append(f"program:{body.program}")
    if body.program == "kyb" and body.subjectKind == "enterprise":
        score += 15
        reasons.append("program:kyb")
    if body.program == "right_to_work":
        score += 10
        reasons.append("program:right_to_work")
    jur = (body.jurisdiction or "GB").upper()
    if jur in HIGH_RISK_JUR:
        score += 35
        flags.append({"code": "high_risk_jurisdiction", "severity": "critical", "source": "rules"})
        reasons.append(f"jurisdiction:{jur}")
    score = min(100, score)
    return {"score": score, "band": _band(score), "flags": flags,
            "reasons": reasons, "model": "verification-compliance-v1"}


@router.post("/score-check")
def score_check(body: CheckIn) -> Dict[str, Any]:
    t = body.checkType
    if t in {"sanctions", "pep"}:
        return {"result": "consider", "score": 0.5,
                "payload": {"note": "deterministic stub — requires human review"}}
    if t == "document":
        return {"result": "clear", "score": 0.85,
                "payload": {"note": "deterministic clear"}}
    return {"result": "consider", "score": 0.6, "payload": {"note": "deterministic"}}


@router.post("/desk-risk")
def desk_risk(body: DeskRiskIn) -> Dict[str, Any]:
    k = body.kpis or {}
    breached = int(k.get("slaBreached") or 0)
    by_band = k.get("casesByBand") or {}
    critical = int(by_band.get("critical") or 0)
    high     = int(by_band.get("high") or 0)
    score = min(100, breached * 8 + critical * 15 + high * 5)
    return {"score": score, "band": _band(score), "model": "vc-desk-risk-v1",
            "factors": {"sla_breached": breached, "critical": critical, "high": high}}
